File: src/core/reconcile.ts
/**
 * Publish-time anchor reconciliation.
 *
 * When a new version is published, an element anchor may no longer resolve (the
 * artifact moved or changed). For each comment alive at the version, we try to
 * re-bind it to the right aid by fingerprint + nearest-heading, appending an
 * `anchor_changed` event. If no confident match exists we mark it `lost` so it
 * renders unanchored rather than silently pointing at the wrong artifact.
 * Older versions keep their own anchors (events are never mutated).
 */
import type {
  Anchor,
  AnchorFallback,
  Comment,
  CommentEvent,
  StampedArtifact,
} from './comment.types.js';
import { appendEvent, ensureMigrated } from './comment-events.js';
import { snapshotAt } from './comment-fold.js';

/** A drifted anchor that may need re-binding. */
type DriftableAnchor = Extract<Anchor, { kind: 'element' | 'lost' }>;

/** Extract the aid an anchor currently targets, if any. */
function knownAid(a: Anchor): string | null {
  if (a.kind === 'element' && a.aid) return a.aid;
  if (a.kind === 'element' && a.selector) {
    return /\[data-tdoc-aid="([\w]+)"\]/.exec(a.selector)?.[1] ?? null;
  }
  return null;
}

/** Find the single confident re-bind target for a drifted anchor, or null. */
function findRebindAid(a: DriftableAnchor, aids: StampedArtifact[]): string | null {
  const wantTag = a.fingerprint?.tag ?? a.label?.toLowerCase() ?? '';
  const wantHead = a.fallback?.nearestHeading?.text;
  const matches = aids.filter(
    (x) =>
      (!wantTag || x.tag === wantTag) &&
      (!wantHead || (x.heading ?? '').toLowerCase() === wantHead.toLowerCase()),
  );
  if (matches.length === 1) return matches[0]!.aid;
  if (matches.length === 0) {
    const tagOnly = aids.filter((x) => !wantTag || x.tag === wantTag);
    if (tagOnly.length === 1) return tagOnly[0]!.aid;
  }
  return null;
}

/** Carry over a drifted anchor's optional fingerprint/fallback metadata. */
function carry(a: DriftableAnchor): { fingerprint?: { tag?: string }; fallback?: AnchorFallback } {
  return {
    ...(a.fingerprint ? { fingerprint: a.fingerprint } : {}),
    ...(a.fallback ? { fallback: a.fallback } : {}),
  };
}

/** The new anchor to record for a drifted comment: a rebind, a lost marker, or none. */
function nextAnchor(a: DriftableAnchor, aids: StampedArtifact[]): Anchor | null {
  const newAid = findRebindAid(a, aids);
  if (newAid) {
    return {
      kind: 'element',
      aid: newAid,
      selector: `[data-tdoc-aid="${newAid}"]`,
      label: a.label ?? a.fingerprint?.tag ?? 'element',
      ...carry(a),
    };
  }
  if (a.kind === 'lost') return null; // already lost, no candidate — don't churn the log
  return {
    kind: 'lost',
    reason: 'no_candidate',
    ...(a.label ? { label: a.label } : {}),
    ...carry(a),
  };
}

/** Build the `anchor_changed` event recording a reconciled anchor. */
function reconcileEvent(anchor: Anchor, version: number, at: string): CommentEvent {
  return {
    kind: 'anchor_changed',
    at_version: version,
    at,
    by: 'reconcile',
    reset_status: false,
    anchor,
  };
}

/** Reconcile a single comment's drifted anchor at `version`, if needed. */
function reconcileComment(
  c: Comment,
  aids: StampedArtifact[],
  byAid: Set<string>,
  version: number,
  at: string,
): void {
  const snap = snapshotAt(c, version);
  if (!snap || snap.deleted) return;
  const a = snap.anchor;
  if (!a || (a.kind !== 'element' && a.kind !== 'lost')) return;

  const aid = knownAid(a);
  if (aid && byAid.has(aid)) return; // still valid

  const anchor = nextAnchor(a, aids);
  if (anchor) appendEvent(c, reconcileEvent(anchor, version, at));
}

/**
 * Reconcile open comment anchors against the freshly-stamped artifact set for a
 * version. Mutates `comments` in place (appending events) and returns it.
 */
export function reconcileAnchors(
  comments: Comment[],
  aidsInVersion: StampedArtifact[],
  v: number,
): Comment[] {
  ensureMigrated(comments);
  const byAid = new Set(aidsInVersion.map((a) => a.aid));
  const version = Number(v) || 1;
  const at = new Date().toISOString();
  for (const c of comments) reconcileComment(c, aidsInVersion, byAid, version, at);
  return comments;
}

File: internal/core/reconcile.go
package core

import (
	"regexp"
	"strings"
	"time"
)

// Publish-time anchor reconciliation, ported from reconcile.ts. When a new
// version is published, an element anchor may no longer resolve. For each comment
// alive at the version we try to re-bind it by fingerprint + nearest-heading,
// appending an anchor_changed event; otherwise we mark it lost.

var aidSelectorRe = regexp.MustCompile(`\[data-odoc-aid="(\w+)"\]`)

// knownAid extracts the aid an anchor currently targets, if any.
func knownAid(a *Anchor) string {
	if a.Kind == "element" && a.AID != "" {
		return a.AID
	}
	if a.Kind == "element" && a.Selector != "" {
		if m := aidSelectorRe.FindStringSubmatch(a.Selector); m != nil {
			return m[1]
		}
	}
	return ""
}

// findRebindAid finds the single confident re-bind target for a drifted anchor.
func findRebindAid(a *Anchor, aids []StampedArtifact) string {
	wantTag := ""
	if a.Fingerprint != nil && a.Fingerprint.Tag != "" {
		wantTag = a.Fingerprint.Tag
	} else if a.Label != "" {
		wantTag = strings.ToLower(a.Label)
	}
	wantHead := ""
	if a.Fallback != nil && a.Fallback.NearestHeading != nil {
		wantHead = a.Fallback.NearestHeading.Text
	}

	var matches []StampedArtifact
	for _, x := range aids {
		tagOK := wantTag == "" || x.Tag == wantTag
		headOK := wantHead == "" || (x.Heading != nil && strings.EqualFold(*x.Heading, wantHead))
		if tagOK && headOK {
			matches = append(matches, x)
		}
	}
	if len(matches) == 1 {
		return matches[0].AID
	}
	if len(matches) == 0 {
		var tagOnly []StampedArtifact
		for _, x := range aids {
			if wantTag == "" || x.Tag == wantTag {
				tagOnly = append(tagOnly, x)
			}
		}
		if len(tagOnly) == 1 {
			return tagOnly[0].AID
		}
	}
	return ""
}

// nextAnchor computes the new anchor to record for a drifted comment: a rebind, a
// lost marker, or nil (no change).
func nextAnchor(a *Anchor, aids []StampedArtifact) *Anchor {
	newAID := findRebindAid(a, aids)
	if newAID != "" {
		label := a.Label
		if label == "" && a.Fingerprint != nil {
			label = a.Fingerprint.Tag
		}
		if label == "" {
			label = "element"
		}
		return &Anchor{
			Kind:        "element",
			AID:         newAID,
			Selector:    `[data-odoc-aid="` + newAID + `"]`,
			Label:       label,
			Fingerprint: a.Fingerprint,
			Fallback:    a.Fallback,
		}
	}
	if a.Kind == "lost" {
		return nil // already lost, no candidate — don't churn the log
	}
	return &Anchor{
		Kind:        "lost",
		Reason:      "no_candidate",
		Label:       a.Label,
		Fingerprint: a.Fingerprint,
		Fallback:    a.Fallback,
	}
}

func reconcileEvent(anchor *Anchor, version int, at string) CommentEvent {
	rs := false
	return CommentEvent{
		Kind: "anchor_changed", AtVersion: version, At: at, By: "reconcile",
		ResetStatus: &rs, Anchor: anchor,
	}
}

func reconcileComment(c *Comment, aids []StampedArtifact, byAID map[string]struct{}, version int, at string) {
	snap := SnapshotAt(c, version)
	if snap == nil || snap.Deleted {
		return
	}
	a := snap.Anchor
	if a == nil || (a.Kind != "element" && a.Kind != "lost") {
		return
	}
	aid := knownAid(a)
	if aid != "" {
		if _, ok := byAID[aid]; ok {
			return // still valid
		}
	}
	if anchor := nextAnchor(a, aids); anchor != nil {
		AppendEvent(c, reconcileEvent(anchor, version, at))
	}
}

// ReconcileAnchors reconciles open comment anchors against the freshly-stamped
// artifact set for a version, mutating comments in place.
func ReconcileAnchors(comments []Comment, aidsInVersion []StampedArtifact, v int) {
	EnsureMigrated(comments)
	byAID := make(map[string]struct{}, len(aidsInVersion))
	for _, a := range aidsInVersion {
		byAID[a.AID] = struct{}{}
	}
	version := v
	if version == 0 {
		version = 1
	}
	at := time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
	for i := range comments {
		reconcileComment(&comments[i], aidsInVersion, byAID, version, at)
	}
}
